src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -479,7 +479,8 @@ public override bool HandleCall(MethodIL callingMethodBody, MethodDesc calledMet
                                         && systemTypeValue.RepresentedType.Type.GetParameterlessConstructor() is MethodDesc ctorMethod
                                         && !reflectionMarker.Factory.MetadataManager.IsReflectionBlocked(ctorMethod))
                                     {
-                                        reflectionMarker.Dependencies.Add(reflectionMarker.Factory.ReflectedMethod(ctorMethod), "Marshal API");
+                                        reflectionMarker.Dependencies.Add(reflectionMarker.Factory.ReflectedType(systemTypeValue.RepresentedType.Type), "Marshal API");
+                                        reflectionMarker.Dependencies.Add(reflectionMarker.Factory.ReflectedMethod(ctorMethod.GetCanonMethodTarget(CanonicalFormKind.Specific)), "Marshal API");
```

Comment:
I'm probably missing something - my understanding of this change here is that we were missing the specific generic parameters on the type because the conversion to canon method target would loose some of it.

But then in the ldtoken case we do the same but only for the generic parameters on the method. I guess what I find confusing is why are the two cases different:
* One makes sure the owning type is properly reflectable (the specific instantiation), but doesn't care about the generic parameters of the method (this is probably because it's always a .ctor and thus doesn't have generic parameters)
* The other makes sure that generic parameters to the method are properly reflectable, but doesn't care about the owning type of the method.

Reply:
`Marshal.PtrToStructure(Type)` is essentially reflection - it will reflection activate this type and run the ctor. We could in theory have a System.Type that doesn't have a full methodtable due to https://github.com/dotnet/runtime/pull/85810#issuecomment-1535929099 and if we were to run into this, this codepath could throw at runtime.

But maybe this is too paranoid now that I look at it again. I'll take it out. I wouldn't be able to come up with a test that fails without this line and we're unlikely to add optimizations that would lead to this being necessary (and if we do, we'll need to revise places like this anyway).

The runtime method handle case (for the buggy case that triggered me to look at this in the first place) handles the owning type because the data structure behind runtime method handle holds a strong reference to the owning type.

Reply:
Thanks for the explanation. So just that I fully understand it - the only reason we don't need to mark method generic parameters in this case is because it's a .ctor so it can't be a generic method, otherwise this would also need to go over them, just like the runtime method handle case, right?

Reply:
> Thanks for the explanation. So just that I fully understand it - the only reason we don't need to mark method generic parameters in this case is because it's a .ctor so it can't be a generic method, otherwise this would also need to go over them, just like the runtime method handle case, right?

Yep.